src/data/repositories/TrainingModuleDefaultRepository.ts

```diff
@@ -298,38 +304,53 @@ export class TrainingModuleDefaultRepository implements TrainingModuleRepository
         });
     }
 
-    private async buildDomainModel(
-        model: PersistedTrainingModule
-    ): Promise<Omit<TrainingModule, "progress" | "outdated" | "builtin">> {
-        if (model._version !== 1) {
-            throw new Error(`Unsupported revision of module: ${model._version}`);
-        }
-
-        const { created, lastUpdated, type, contents, ...rest } = model;
-        const validType = isValidTrainingType(type) ? type : "app";
+    private async buildDomainModels(
+        models: PersistedTrainingModule[]
+    ): Promise<Omit<TrainingModule, "progress" | "outdated" | "builtin">[]> {
         const currentUser = await this.configRepository.getUser();
         const instanceVersion = await this.instanceRepository.getVersion();
 
-        return {
-            ...rest,
-            contents: {
-                ...contents,
-                steps: contents.steps.map((step, stepIdx) => ({
-                    ...step,
-                    id: `${model.id}-step-${stepIdx}`,
-                    pages: step.pages.map((page, pageIdx) => ({
-                        ...page,
-                        id: `${model.id}-page-${stepIdx}-${pageIdx}`,
+        return promiseMap(models, async model => {
+            if (model._version !== 1) {
+                throw new Error(`Unsupported revision of module: ${model._version}`);
+            }
+
+            const { created, lastUpdated, type, contents, ...rest } = model;
+            const validType = isValidTrainingType(type) ? type : "app";
+
+            const checkPagePermissions = (page: TrainingModulePageOptionalPermissions, permission: "read" | "write") =>
+                validateUserPagePermissions({
+                    module: model,
+                    item: page,
+                    permission,
+                    currentUser,
+                });
+
+            return {
+                ...rest,
+                contents: {
+                    ...contents,
+                    steps: contents.steps.map((step, stepIdx) => ({
+                        ...step,
+                        id: `${model.id}-step-${stepIdx}`,
```

Comment:
I think it would be good to extract this id generation to a function because we have some similar logic on the `TrainingModuleHelpers.ts` file.